Low battery mode: make client-side configurable

garvankeeley opened the pull request:
Moved the low battery check in active scanning to the client side, and added a developer setting to set the min battery level.

garvankeeley committed:
adding support for low battery mode in the client side, which can override default service-side behaviour.
TODO actually take action on low battery.

garvankeeley committed:
Remove receiver in manifest for low battery notifications. The code
already checks this as part of scanning. Also, it is inflexible as it
notfies on a set battery level -instead, the code should handle a
parameterized battery level.

garvankeeley committed:
Send turn off broadcast when battery is low

cascheberg commented:
So I will close #1250 unmerged and file a new issue about user notification.

Some comments about this PR (vs. #1250):
- the user should be able to ignore low battery and restart scans, so the app should sent the turn off intent only once
- you could send a dedicated low battery intent instead of the `INTENT_TURN_OFF` to prepare for user notifications, I used the `notificationDrawerEventReceiver` for that
- in my version I hooked up to the `ACTION_BATTERY_CHANGED` intent instead of creating my own runnable
- I kept the `TurnOffReceiver` because I thought it would help with testing

garvankeeley commented:
Thanks for the review, good points made, I am updating my PR to address most of these,

garvankeeley committed:
Update with these changes:
- System ACTION_BATTERY_CHANGED intent drives the battery check code
- In the ClientStumblerService, in the battery checking code, if the
  battery is low and the user starts the service manually, battery
checking is deactivated until the battery is ok again.
- Send a dedicated low battery intent from the ClientStumblerService on
  low battery. (Received by MainApp)

garvankeeley commented:
@cascheberg Updated based on your feedback
@crankycoder this needs an automated test. But I don't think we have the framework in place to do the testing I want, yet.

Djfe commented:
Some of my ideas:
It should allow 0% battery level as the stopping point -> stop scanning when the device is shutting down, to save the scanned data before it gets lost

Is there a check at the first scan start? it shouldn't start scanning automatically on app start on low battery. Or is it doing that already because of a general check?

EDIT:
what do you think about using a picker for the battery percentage instead?
http://developer.android.com/design/building-blocks/pickers.html

garvankeeley commented:
I can add 0% .

The checks begin after scanning starts when the battery service reports a change.

crankycoder commented on android/src/main/java/org/mozilla/mozstumbler/service/stumblerthread/scanners/ScanManager.java:
```diff
@@ -32,28 +32,33 @@
+            if (mPassiveModeBatteryState == PassiveModeBatteryState.IGNORE_BATTERY_STATE) {
+                return;
+            }
+            final int kMinBatteryPct = 15;
```
Shouldn't this use the battery preference settings instead of just 15?

garvankeeley replied:
This is hard coded for out-of-the-box Fennec use.

garvankeeley committed:
Added 0% battery option

cascheberg commented:
There have been lots of fixes/changes since the last release, so it would be nice to have another official pre-release. Do you have any plans regarding that?

garvankeeley commented:
Yeah, hoping to do a github release tomorrow. Then maybe a week later do an f-droid and playstore release if that build is looking healthy.
